cmd/tpm-manager/tpm-manager.go

```diff
@@ -0,0 +1,506 @@
+			pcrdata[pcr] = pcrtmp
+		}
+		manager.pcrconfigs = append(manager.pcrconfigs, pcrdata)
+	} else if manager.pcrconfigdir != "" {
+		err = filepath.Walk(manager.pcrconfigdir, func(path string, f os.FileInfo, err error) error {
+			if f.IsDir() {
+				return nil
+			}
+			pcrconfig, err := ioutil.ReadFile(path)
+			if err != nil {
+				glog.Errorf("Unable to read PCR configuration %s: %v", path, err)
+				return err
+			}
+			pcrdata := make(map[string]tpm.PCRConfig)
+			err = json.Unmarshal(pcrconfig, &pcrdata)
+			if err != nil {
+				glog.Errorf("Unable to parse valid PCR configuration %s: %v", path, err)
+				return err
+			}
+			for pcr, _ := range pcrdata {
+				pcrtmp := pcrdata[pcr]
+				pcrtmp.Source = path
+				pcrdata[pcr] = pcrtmp
+			}
+			manager.pcrconfigs = append(manager.pcrconfigs, pcrdata)
+			return nil
+		})
+	} else {
+		manager.pcrconfigs, err = manager.tpmhandler.GetPolicies()
+		if err != nil {
+			glog.Errorf("Unable to obtain PCR configuration: %v", err)
+			return nil
+		}
+	}
+	return nil
+}
+
+func verifyNode(node *api.Node) (err error) {
+	address, err := nodeutil.GetNodeHostIP(node)
+	if err != nil {
+		return err
+	}
+	host := fmt.Sprintf("%s:23179", address.String())
+	tpmdata, err := manager.tpmhandler.Get(host, manager.allowunknown)
+	if err != nil {
+		glog.Errorf("Unable to obtain TPM data for node %s: %v", address.String(), err)
+		invalidateNode(node)
+		return nil
+	}
+	quote, log, err := tpm.Quote(tpmdata)
+	if err != nil {
+		glog.Errorf("Unable to obtain TPM quote for node %s: %v", address.String(), err)
+		invalidateNode(node)
+		return nil
+	}
+
+	err = tpm.ValidateLog(log, quote)
+	if err != nil {
+		glog.Errorf("TPM event log does not match quote for node %s", address.String())
+		invalidateNode(node)
+		return nil
+	}
+
+	logstate, err := tpm.ValidatePCRs(log, quote, manager.pcrconfigs)
+	jsonlog, _ := json.Marshal(logstate)
+	if node.ObjectMeta.Annotations == nil {
+		node.ObjectMeta.Annotations = make(map[string]string)
+	}
+	node.ObjectMeta.Annotations["tpm.coreos.com/logstate"] = string(jsonlog)
+	if err != nil {
+		glog.Errorf("TPM quote PCRs don't validate for node %s", address.String())
+		invalidateNode(node)
+		return nil
+	}
+
+	glog.Errorf("TPM quote for node %s validates", address.String())
+	validateNode(node)
+	return nil
+}
+
+func verifyAndUpdate(node *api.Node) {
+	verifyNode(node)
+
+	// The state that the node will be updated to
+	newstate := isTrusted(node)
+	newnode, err := manager.client.Nodes().Get(node.Name)
+	if err != nil {
+		glog.Errorf("Unable to obtain node state for %s: %v", node.Name, err)
+		return
+	}
+	if newnode.ObjectMeta.Annotations == nil {
+		newnode.ObjectMeta.Annotations = make(map[string]string)
+	}
+	currenttime := time.Now().Unix()
+
+	// If we're transitioning state, update the metadata
+	if isTrusted(newnode) != newstate {
+		if newstate == false {
+			newnode.ObjectMeta.Annotations["tpm.coreos.com/untrustedsince"] = strconv.FormatInt(currenttime, 10)
```

Comment:
Can we add consts for all these key strings. Also, should these be version namespaced? e.g. `tpm.alpha.coreos.com`?